classes/observers/traits/UsageEvent.inc.php

```
@@ -0,0 +1,164 @@
+            case Application::ASSOC_TYPE_SUBMISSION_FILE:
+                $canonicalUrlOp = 'download';
+                $canonicalUrlParams = [$submission->getId()];
+                $router = $this->request->getRouter(); /** @var PageRouter $router */
+                $op = $router->getRequestedOp($this->request);
+                $args = $router->getRequestedArgs($this->request);
+                if ($op == 'download' && count($args) > 1) {
+                    if ($args[1] == 'version' && count($args) == 5) {
+                        $publicationId = (int) $args[2];
+                        $canonicalUrlParams[] = 'version';
+                        $canonicalUrlParams[] = $publicationId;
+                    }
+                }
+                $canonicalUrlParams[] = $representation->getId();
+                $canonicalUrlParams[] = $submissionFile->getId();
+                break;
+            case Application::ASSOC_TYPE_SUBMISSION:
+                $canonicalUrlOp = 'view';
+                if ($application->getName() == 'omp') {
+                    $canonicalUrlOp = 'book';
+                }
+                $canonicalUrlParams = [$submission->getId()];
+                $router = $this->request->getRouter(); /** @var PageRouter $router */
+                $op = $router->getRequestedOp($this->request);
+                $args = $router->getRequestedArgs($this->request);
+                if ($op == $canonicalUrlOp && count($args) > 1) {
+                    if ($args[1] == 'version' && count($args) == 3) {
+                        $publicationId = (int) $args[2];
+                        $canonicalUrlParams[] = 'version';
+                        $canonicalUrlParams[] = $publicationId;
+                    }
+                }
+                break;
+            case Application::getContextAssocType():
+                $canonicalUrlOp = '';
+                $canonicalUrlPage = 'index';
+                break;
+        }
```

This should throw an exception in the default case.

---

I needed to reorganize the code a little bit for this...